### PR TITLE
[FAD-40] Metamask Connect Button

### PR DESCRIPTION
**Pull request by sachasmart-weavik**
### Description

> Metamask connect button with various states. Metamask/Web3-react does not allow the app to disconnect the user. This functionality doesn't work in Goldfinch either. 

Ref: https://docs.cloud.coinbase.com/wallet-sdk/docs/web3-react#i-want-to-give-the-user-the-option-to-disconnect-from-my-dapp

Submitting as draft because I changed up a lot of the Goldfinch conventions as this way makes more sense/feels more familiar. 

[FAD-40]

### Checklist

- [ ] `CHANGELOG` has been updated (_if appropriate_)
- [ ] Environment variables updated (_if appropriate_)
ezgif com-gif-maker (2)

[FAD-40]: https://weavik.atlassian.net/browse/FAD-40?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sachasmart-weavik** [comment]
A few updates based on Vineeth's input. Mainly reducing the amount of `useState`. 
ezgif com-gif-maker (3)

File: packages/client3/components/nav/wallet-button.tsx [hunk 1]
```diff
@@ -1,43 +1,119 @@
...
+    const provider = await detectEthereumProvider();
+    if (provider) {
+      return true;
+    }
+    return false;
+  }, []);
+
+  const handleWalletConnect = async () => {
+    setIsLoading(true);
+    if (await detectBrowser) {
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:31]: Should this be 
```suggestion
    if (await detectBrowser()) {
```

File: packages/client3/components/nav/wallet-button.tsx [hunk 2]
```diff
@@ -1,43 +1,119 @@
...
-          {account.substring(0, 6)}...{account.substring(account.length - 4)}
-        </span>
-        <Identicon account={account} scale={3} />
-      </Button>
-    </Popover>
+    }
+    setIsLoading(false);
+  };
+
+  function handleAddressFormat(account: string) {
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:39]: Would recommend extracting this function to a utils class to provide more reusability and also would recommend using arrow function

File: packages/client3/components/nav/wallet-button.tsx [hunk 3]
```diff
@@ -1,43 +1,119 @@
...
+        state: handleAddressFormat(account as string),
+      };
+    } else {
+      return {
+        icon: "",
+        type: ButtonType.PRIMARY,
+        state: ButtonStateText.CONNECT,
+      };
+    }
+  }, [account, error, isActive, isLoading]);
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:80]: Sorry, I wasn't clear enough. You could avoid using useCallback in this function call as this function isn't being passed as a dependency or passed to a component.
Also,  a minor thing would be to remove `else` as there is already return in the if blocks.

This style 
```
{
        icon: "",
        type: ButtonType.PRIMARY,
        state: ButtonStateText.CONNECT,
      } 
```
seems to be returned in two conditionals, final return would be enough i think

File: packages/client3/components/nav/wallet-button.tsx [hunk 4]
```diff
@@ -1,43 +1,119 @@
...
+
+  const buttonStyles = getButtonStyles();
+
+  const { data } = useCurrentUserWalletInfoQuery({
+    variables: { userAccount: "" },
+    fetchPolicy: "network-only",
+  });
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:87]: It's a general practice to keep all the hooks together at the begining

File: packages/client3/components/nav/wallet-button.tsx [hunk 5]
```diff
@@ -1,43 +1,119 @@
...
+  const user = data?.user;
+
+  useEffect(() => {
+    if (user) {
+      getAccount(user);
+    }
+  }, [getAccount, user]);
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:94]: Same as above

File: packages/client3/components/nav/wallet-button.tsx [hunk 6]
```diff
@@ -1,43 +1,119 @@
...
+
+  return buttonStyles.tooltip ? (
+    <>
```
Comment (v1n33th) [inline packages/client3/components/nav/wallet-button.tsx:97]: I think it's not needed

**Review by v1n33th — CHANGES_REQUESTED** [review]

File: packages/client3/components/nav/wallet-button.tsx (hunk 3 above)
Comment (sachasmart-weavik) [reply packages/client3/components/nav/wallet-button.tsx:80]: My bad. isLoading should've had a connecting state ("Processing....")
```
    if (isLoading) {
      return {
        icon: "",
        type: ButtonType.PRIMARY,
        state: ButtonStateText.CONNECTING,
      };
    }
```

**Review by v1n33th — APPROVED** [review]

**Pull request merged**